### PR TITLE
Provide optional stream to copy_to_external API. Fix sync issue

### PR DESCRIPTION
**Pull request by jantonguirao**
Signed-off-by: Joaquin Anton <janton@nvidia.com>

### CODE REVIEW

**Comment by jantonguirao** [comment]
!build

File: dali/c_api/c_api.cc [hunk 1]
```diff
@@ -203,7 +203,11 @@ unsigned daliGetNumOutput(daliPipelineHandle* pipe_handle) {
 template <typename T>
 static void daliCopyTensorListNToHelper(dali::DeviceWorkspace* ws, void* dst, int n,
                                         device_type_t dst_type) {
-  dali::CopyToExternalTensor(&(ws->Output<T>(n)), dst, (dali::device_type_t)dst_type);
+  dali::CopyToExternalTensor(
+    &(ws->Output<T>(n)),
+    dst,
+    (dali::device_type_t)dst_type,
+    (cudaStream_t) 0);
```
Comment (JanuszL) [inline dali/c_api/c_api.cc]: Bad idea

File: dali/c_api/c_api.cc [hunk 2]
```diff
@@ -222,7 +226,7 @@ static void daliCopyTensorNToHelper(dali::DeviceWorkspace* ws, void* dst, int n,
                                     device_type_t dst_type) {
   dali::Tensor<T> t;
   t.ShareData(&(ws->Output<T>(n)));
-  dali::CopyToExternalTensor(t, dst, (dali::device_type_t)dst_type);
+  dali::CopyToExternalTensor(t, dst, (dali::device_type_t)dst_type, (cudaStream_t) 0);
```
Comment (JanuszL) [inline dali/c_api/c_api.cc]: Bad idea

File: dali/c_api/c_api.cc (hunk 1 above)
Comment (Kh4L) [inline dali/c_api/c_api.cc]: IMO the stream should be given as an arg of `daliCopyTensorListNToHelper` and TF should decide what to do about it

File: dali/plugin/copy.cu [hunk 3]
```diff
@@ -23,12 +23,18 @@
...
+                                       cudaStream_t stream) {}
 
 template <>
-void CopyToExternalTensorHelper<CPUBackend>(const dali::Buffer<CPUBackend> &src, void *dst,
-                                            device_type_t dst_type, size_t num) {
+void CopyToExternalTensorHelper<CPUBackend>(const dali::Buffer<CPUBackend> &src,
+                                            void *dst,
+                                            device_type_t dst_type,
+                                            size_t num,
+                                            cudaStream_t stream) {
```
Comment (JanuszL) [inline dali/plugin/copy.cu]:
```suggestion
                                            cudaStream_t /*stream*/) {
```

File: dali/c_api/c_api.cc (hunk 1 above)
Comment (jantonguirao) [reply dali/c_api/c_api.cc]: done

File: dali/c_api/c_api.cc (hunk 2 above)
Comment (jantonguirao) [reply dali/c_api/c_api.cc]: fixed

File: dali/c_api/c_api.cc (hunk 1 above)
Comment (jantonguirao) [reply dali/c_api/c_api.cc]: fixed

File: dali/plugin/copy.cu (hunk 3 above)
Comment (jantonguirao) [reply dali/plugin/copy.cu]: fixed

**Comment by jantonguirao** [comment]
!build

**Comment by Kh4L** [comment]
What about adding an `non_blocking` option in which we would skip https://github.com/NVIDIA/DALI/blob/3ebee76d89fc698387ae6d40b70f9ab2a2be20bb/dali/plugin/copy.cu#L65 ?

**Comment by Kh4L** [comment]
> What about adding an `non_blocking` option in which we would skip
> 
> [DALI/dali/plugin/copy.cu](https://github.com/NVIDIA/DALI/blob/3ebee76d89fc698387ae6d40b70f9ab2a2be20bb/dali/plugin/copy.cu#L65)
> 
> Line 65 in [3ebee76](/NVIDIA/DALI/commit/3ebee76d89fc698387ae6d40b70f9ab2a2be20bb)
>  CUDA_CALL(cudaStreamSynchronize(stream)); 
> ?

Ok so this would be the topic of another PR

**Pull request merged**